File: code-checker/src/main/java/org/example/code/checker/checker/markdown/task/checker/structrue/title/TitleChecker.java
package org.example.code.checker.checker.markdown.task.checker.structrue.title;

import org.example.code.checker.checker.CheckError;
import org.example.code.checker.checker.markdown.domain.StdNode;
import org.example.code.checker.checker.markdown.domain.standard.StandardNodeType;
import org.example.code.checker.checker.markdown.domain.standard.block.Document;
import org.example.code.checker.checker.markdown.domain.standard.block.Heading;
import org.example.code.checker.checker.markdown.domain.standard.block.Paragraph;
import org.example.code.checker.checker.markdown.domain.standard.inline.CodeSpan;
import org.example.code.checker.checker.markdown.domain.standard.inline.Text;
import org.example.flow.engine.node.TaskData;
import org.example.flow.engine.node.TaskDataUtils;
import org.example.flow.engine.node.TaskNode;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * Check title structure and completeness based on {@link Document} domain node.
 * <p>
 * Rules:
 * <ul>
 *     <li>Document must contain a level-1 heading as the title.</li>
 *     <li>The first non-empty paragraph after the title is treated as the title
 *     description and must be present.</li>
 * </ul>
 * <p>
 * On success this task returns {@code TaskData<TitleInfo>} with parsed title info.
 * On failure it returns {@code TaskData<CheckError>} describing the problem.
 */
public class TitleChecker extends TaskNode {

    @Override
    public TaskData<?> task(Map<String, TaskData<?>> input) {
        if (input == null || input.isEmpty()) {
            throw new IllegalArgumentException("input is null or empty");
        }

        Document document =
            TaskDataUtils.getPayload(input, "document", Document.class);
        String fileId = TaskDataUtils.getPayload(input, "fileId", String.class);

        if (document == null) {
            CheckError error = new CheckError(
                "Title.DocumentMissing",
                "Document domain node is missing in input",
                CheckError.Severity.ERROR,
                fileId,
                null,
                null,
                "Title"
            );
            return new TaskData<>(
                CheckError.class,
                TitleChecker.class.getSimpleName(),
                System.currentTimeMillis(),
                error
            );
        }

        Heading h1 = findFirstH1(document);
        if (h1 == null) {
            CheckError error = new CheckError(
                "Title.MissingHeading",
                "Level-1 heading (title) is missing in document",
                CheckError.Severity.ERROR,
                fileId,
                document.getRange(),
                document.getNodeId(),
                "Title"
            );
            return new TaskData<>(
                CheckError.class,
                TitleChecker.class.getSimpleName(),
                System.currentTimeMillis(),
                error
            );
        }

        String titleText = collectInlineText(h1);
        if (titleText == null || titleText.trim().isEmpty()) {
            CheckError error = new CheckError(
                "Title.EmptyHeadingText",
                "Level-1 heading (title) has empty text content",
                CheckError.Severity.ERROR,
                fileId,
                h1.getRange(),
                h1.getNodeId(),
                "Title"
            );
            return new TaskData<>(
                CheckError.class,
                TitleChecker.class.getSimpleName(),
                System.currentTimeMillis(),
                error
            );
        }

        Paragraph descParagraph = findFirstParagraphAfter(document, h1);
        if (descParagraph == null) {
            CheckError error = new CheckError(
                "Title.MissingDescription",
                "Title description paragraph after H1 is missing",
                CheckError.Severity.ERROR,
                fileId,
                document.getRange(),
                document.getNodeId(),
                "Title"
            );
            return new TaskData<>(
                CheckError.class,
                TitleChecker.class.getSimpleName(),
                System.currentTimeMillis(),
                error
            );
        }

        String descriptionText = collectInlineText(descParagraph);
        if (descriptionText == null || descriptionText.trim().isEmpty()) {
            CheckError error = new CheckError(
                "Title.EmptyDescription",
                "Title description paragraph after H1 is empty",
                CheckError.Severity.ERROR,
                fileId,
                descParagraph.getRange(),
                descParagraph.getNodeId(),
                "Title"
            );
            return new TaskData<>(
                CheckError.class,
                TitleChecker.class.getSimpleName(),
                System.currentTimeMillis(),
                error
            );
        }

        List<String> extraParagraphs = collectExtraParagraphs(document, descParagraph);

        TitleInfo info = new TitleInfo(
            titleText.trim(),
            descriptionText.trim(),
            extraParagraphs
        );

        return new TaskData<>(
            TitleInfo.class,
            TitleChecker.class.getSimpleName(),
            System.currentTimeMillis(),
            info
        );
    }

    private Heading findFirstH1(Document document) {
        List<StdNode> children = document.getChildren();
        if (children == null || children.isEmpty()) {
            return null;
        }
        for (StdNode child : children) {
            if (child instanceof Heading) {
                Heading h = (Heading) child;
                if (h.getLevel() == 1) {
                    return h;
                }
            }
        }
        return null;
    }

    private Paragraph findFirstParagraphAfter(Document document, Heading h1) {
        List<StdNode> children = document.getChildren();
        if (children == null || children.isEmpty() || h1 == null) {
            return null;
        }
        boolean seenH1 = false;
        for (StdNode child : children) {
            if (child == h1) {
                seenH1 = true;
                continue;
            }
            if (!seenH1) {
                continue;
            }
            if (child instanceof Paragraph) {
                return (Paragraph) child;
            }
        }
        return null;
    }

    /**
     * Collect optional extra paragraphs that appear after the main title description.
     * These are consecutive {@link Paragraph} blocks following the first description
     * paragraph, and stop when a non-paragraph block is encountered.
     */
    private List<String> collectExtraParagraphs(Document document, Paragraph descParagraph) {
        List<String> result = new ArrayList<>();
        List<StdNode> children = document.getChildren();
        if (children == null || children.isEmpty() || descParagraph == null) {
            return result;
        }

        boolean seenDesc = false;
        for (StdNode child : children) {
            if (child == descParagraph) {
                seenDesc = true;
                continue;
            }
            if (!seenDesc) {
                continue;
            }
            if (child instanceof Paragraph) {
                String text = collectInlineText(child);
                if (text != null && !text.trim().isEmpty()) {
                    result.add(text.trim());
                }
            } else {
                // Stop when encountering the first non-paragraph block
                break;
            }
        }

        return result;
    }

    private String collectInlineText(StdNode node) {
        if (node == null || !node.hasChildren()) {
            return "";
        }
        StringBuilder sb = new StringBuilder();
        for (StdNode child : node.getChildren()) {
            appendInlineText(child, sb);
        }
        return sb.toString().trim();
    }

    private void appendInlineText(StdNode node, StringBuilder out) {
        if (node == null) {
            return;
        }
        if (node.getNodeType() == StandardNodeType.TEXT && node instanceof Text) {
            out.append(((Text) node).getContent());
        } else if (node.getNodeType() == StandardNodeType.CODE_SPAN && node instanceof CodeSpan) {
            out.append(((CodeSpan) node).getCode());
        } else if (node.hasChildren()) {
            for (StdNode child : node.getChildren()) {
                appendInlineText(child, out);
            }
        }
    }
}

File: code-checker/src/main/kotlin/org/example/code/checker/checker/markdown/task/checker/structrue/title/TitleChecker.kt
package org.example.code.checker.checker.markdown.task.checker.structrue.title

import org.example.code.checker.checker.common.CheckError
import org.example.code.checker.checker.markdown.domain.MdDomain
import org.example.code.checker.checker.markdown.domain.standard.block.Heading
import org.example.code.checker.checker.markdown.domain.standard.block.Paragraph
import org.example.code.checker.checker.markdown.domain.standard.inline.CodeSpan
import org.example.code.checker.checker.markdown.domain.standard.inline.Text
import org.example.code.checker.checker.utils.TreeNode
import org.example.flow.engine.node.TaskData
import org.example.flow.engine.node.TaskDataUtils
import org.example.flow.engine.node.TaskNode

/**
 * Title structure checking rule.
 *
 * This node validates that the "title section" of a document is complete.
 * For a Terraform-style resource document such as `huaweicloud_cdn_ip_information.md`,
 * the title section is defined as:
 *
 * - Starts from the first level‑1 heading (`# ...`).
 * - Ends right before the first level‑2 heading (`## ...`).
 * - Lines starting with `->` inside this section are optional tips and are not required.
 *
 * Rules checked:
 * - A level‑1 heading must exist in the document.
 * - Between the first H1 and the first H2, there must be at least one non‑empty
 *   paragraph, which is treated as the title description.
 *
 * Input:
 * - workingDocument: TaskData<TreeNode<MdDomain>> – document tree after
 *   FrontMatter has been consumed.
 * - fileId:          TaskData<String>              – file identifier or path.
 *
 * Output:
 * - TaskData<List<CheckError>> – all title-related errors (empty if title is complete).
 */
class TitleChecker : TaskNode() {

    override fun task(input: Map<String, TaskData<*>>?): List<TaskData<*>> {
        if (input == null || input.isEmpty()) {
            throw IllegalArgumentException("input is null or empty")
        }

        @Suppress("UNCHECKED_CAST")
        val document: TreeNode<MdDomain> =
            TaskDataUtils.getPayload(input, "workingDocument", TreeNode::class.java) as TreeNode<MdDomain>
        val fileId: String =
            TaskDataUtils.getPayload(input, "fileId", String::class.java)

        val errors = mutableListOf<CheckError>()

        // 1. Find first H1 heading using TreeNode.Query for type-safe filtering.
        val h1Opt = document
            .query()
            .all()
            .filter(Heading::class.java) { it.level == 1 }
            .first()

        if (!h1Opt.isPresent) {
            errors += CheckError.builder()
                .ruleId("Title.MissingH1")
                .message("Level-1 heading (title) is missing in document")
                .severity(CheckError.Severity.ERROR)
                .fileId(fileId)
                .range(document.data.range)
                .nodeId(null)
                .nodeType("Title")
                .build()
        } else {
            val h1Node: TreeNode<MdDomain> = h1Opt.get()

            // Flatten nodes in pre-order to reason about relative positions.
            val allNodes: List<TreeNode<MdDomain>> = document.query().all().list()

            val h1Index = allNodes.indexOf(h1Node)

            // 2. Find first H2 heading after this H1; if not found, the title section
            //    extends to the end of the document.
            var h2Index = -1
            for (i in (h1Index + 1) until allNodes.size) {
                val data = allNodes[i].data
                if (data is Heading && data.level == 2) {
                    h2Index = i
                    break
                }
            }

            val sectionEndExclusive =
                if (h2Index >= 0) h2Index else allNodes.size

            // 3. Search for the first non-empty paragraph between H1 and the first H2.
            var foundDescription = false
            for (i in (h1Index + 1) until sectionEndExclusive) {
                val node = allNodes[i]
                val data = node.data
                if (data is Paragraph) {
                    val text = collectInlineText(node).trim()
                    if (text.isNotEmpty()) {
                        foundDescription = true
                        break
                    }
                }
            }

            if (!foundDescription) {
                errors += CheckError.builder()
                    .ruleId("Title.MissingDescription")
                    .message("Title description paragraph after H1 is missing before the first H2")
                    .severity(CheckError.Severity.ERROR)
                    .fileId(fileId)
                    .range(h1Node.data.range)
                    .nodeId(null)
                    .nodeType("Title")
                    .build()
            }
        }

        val errorsData = TaskData(
            List::class.java,
            TitleChecker::class.java.simpleName,
            System.currentTimeMillis(),
            errors.toList()
        )

        return listOf(errorsData)
    }

    /**
     * Collect inline text from a paragraph (or any inline container) node.
     * Text content and inline code spans are concatenated in reading order.
     */
    private fun collectInlineText(node: TreeNode<MdDomain>): String {
        val sb = StringBuilder()
        appendInlineText(node, sb)
        return sb.toString()
    }

    private fun appendInlineText(node: TreeNode<MdDomain>?, out: StringBuilder) {
        if (node == null) return

        val data = node.data
        when (data) {
            is Text -> out.append(data.content)
            is CodeSpan -> out.append(data.code)
            else -> {
                val children = node.children ?: return
                for (child in children) {
                    appendInlineText(child, out)
                }
            }
        }
    }
}


